xds/src/main/java/io/grpc/xds/XdsLrsClient.java

```diff
@@ -0,0 +1,325 @@
+import com.google.common.base.Stopwatch;
+import com.google.common.base.Supplier;
+import com.google.protobuf.Duration;
+import com.google.protobuf.Struct;
+import com.google.protobuf.Value;
+import io.envoyproxy.envoy.api.v2.core.Node;
+import io.envoyproxy.envoy.api.v2.endpoint.ClusterStats;
+import io.envoyproxy.envoy.service.load_stats.v2.LoadReportingServiceGrpc;
+import io.envoyproxy.envoy.service.load_stats.v2.LoadStatsRequest;
+import io.envoyproxy.envoy.service.load_stats.v2.LoadStatsResponse;
+import io.grpc.ChannelLogger;
+import io.grpc.ChannelLogger.ChannelLogLevel;
+import io.grpc.LoadBalancer.Helper;
+import io.grpc.ManagedChannel;
+import io.grpc.Status;
+import io.grpc.SynchronizationContext;
+import io.grpc.SynchronizationContext.ScheduledHandle;
+import io.grpc.internal.BackoffPolicy;
+import io.grpc.stub.StreamObserver;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * Client of XDS load reporting service. It should be run in some serialized context.
+ */
+@NotThreadSafe
+final class XdsLrsClient {
+
+  @VisibleForTesting
+  static final String TRAFFICDIRECTOR_GRPC_HOSTNAME
+      = "com.googleapis.trafficdirector.grpc_hostname";
+  private final String serviceName;
+  private final ManagedChannel channel;
+  private final Helper helper;
+  private final SynchronizationContext syncContext;
+  private final ScheduledExecutorService timerService;
+  private final Supplier<Stopwatch> stopwatchSupplier;
+  private final Stopwatch stopwatch;
+  private final ChannelLogger logger;
+  private final BackoffPolicy.Provider backoffPolicyProvider;
+
+  @Nullable
+  private BackoffPolicy lrsRpcRetryPolicy;
+  @Nullable
+  private ScheduledHandle lrsRpcRetryTimer;
+
+  @Nullable
+  private LrsStream lrsStream;
+
+  XdsLrsClient(ManagedChannel channel, Helper helper, Supplier<Stopwatch> stopwatchSupplier,
+      BackoffPolicy.Provider backoffPolicyProvider) {
+    this.channel = checkNotNull(channel, "channel");
+    this.helper = checkNotNull(helper, "helper");
+    this.serviceName = checkNotNull(helper.getAuthority(), "serviceName");
+    this.syncContext = checkNotNull(helper.getSynchronizationContext(), "syncContext");
+    this.stopwatchSupplier = checkNotNull(stopwatchSupplier, "stopwatchSupplier");
+    this.stopwatch = stopwatchSupplier.get();
+    this.logger = checkNotNull(helper.getChannelLogger(), "logger");
+    this.timerService = checkNotNull(helper.getScheduledExecutorService(), "timeService");
+    this.backoffPolicyProvider = checkNotNull(backoffPolicyProvider, "backoffPolicyProvider");
+  }
+
+  // TODO (chengyuanzhang): consider putting to a util class.
+  @VisibleForTesting
+  static long durationToNum(Duration duration, TimeUnit unit) {
+    return unit.convert(duration.getSeconds() * 1000_000_000L + duration.getNanos(),
```

Comment:
`com.google.protobuf.util.Durations#toNanos()` does what you need.

Reply:
ACK.